configuration/device-tree.md

```diff
@@ -199,7 +177,7 @@ A DT overlay comprises a number of fragments, each of which targets one node and
 };
 ```
 
-The `compatible` string identifies this as being for BCM2708, which is the base architecture of the BCM2835 part. For the BCM2836 part you could use a compatible string of "brcm,bcm2709", but unless you are targeting features of the ARM CPUs, the two architectures should be equivalent, so sticking to "brcm,bcm2708" is reasonable. Then comes the first (and in this case only) fragment. Fragments are numbered sequentially from zero. Failure to adhere to this may cause some or all of your fragments to be missed.
+The `compatible` string identifies this as being for BCM2835, which is the base architecture for the Raspberry Pi SoCs; if the overlay makes use of features of a Pi 4 then `brcm,bcm2711` is the correct value to use, otherwise `brcm,bcm2835` can be used for all Pi overlays. Then comes the first (and in this case only) fragment. Fragments should be numbered sequentially from zero. Failure to adhere to this may cause some or all of your fragments to be missed.
```

Comment:
Is the compatible string in the overlay matched on now by the firmware/libdft? That'd be very useful in merging the vc4-kms-v3d and vc4-kms-v3d-pi4 overlays.

Reply:
Not at present, no. How were you thinking it could work?

Reply:
Are "All Pis" and "Pi 4 only" the only (sesnsible) options, are are there other variants?
Is there any way to indicate "compatible with a Pi1 but not a Pi4" (or "Pi2 but not Pi3"), or is that a nonsensical question?
(I'm only playing devil's advocate here :wink: )

Reply:
For now it's just those two options.

Reply:
It was more the wording you'd used that made me think it was working already.

I don't know how it would work within the firmware, but the hardware config between Pi0-3 and Pi4 means the overlays are slightly different, eg enabling more pixelvalves, ddc0/dcc1 instead of i2c2, hdmi0 & hdmi1 instead of hdmi, etc.

It would be nice if we could do a platform agnostic overlay as ```vc4-kms-v3d``` that worked on all platforms, instead of having to use config.txt conditionals.
```
[pi4]
dtoverlay=vc4-kms-v3d-pi4
[!pi4]
dtoverlay=vc4-kms-v3d
[all]
```
if indeed that can work.

We have loaded the base DT, so does /compatible contain a string we can compare against before applying the overlay? Looking in /proc/device-tree/compatible on a running system it appears to.

Reply:
Apart from the compatible strings for device nodes, there is only the top-level node.

There are a number of problems to solve:
1. Fragments that are in legal on all platforms but that we only want to be applied on some platforms. This could be perhaps solved without adding much infrastructure using reserved parameter names that enable and disable fragments.
2. References to labels that are only exist on some platforms. Since symbols are all lumped together without regard to the referencing fragment, solving this one is more tricky - we'd have to delay symbol resolution to the point at which each fragment is applied, and filter out the unwanted fragments in the dtoverlay utility.

Reply:
OK, problem for another day.
For now it might be easier to have a magic string match in the firmware and replace vc4-kms-v3d with vc4-kms-v3d-pi4 if we're on 2711.

Reply:
Agreed.